### PR TITLE
fix: Error if after the migration there are `not null` columns that are not part of the new schema

### PR DESCRIPTION
**Pull request by erezrokah**
#### Summary

Fixes https://github.com/cloudquery/cloudquery/issues/6248 (implements ~~first~~ second solution from the issue)

### CODE REVIEW

File: plugins/destination/postgresql/client/client_test.go [hunk 1]
```diff
@@ -24,3 +32,44 @@ func TestPgPlugin(t *testing.T) {
 		},
 		destination.PluginTestSuiteTests{})
 }
+
+func TestPgPluginPrimaryKeyRename(t *testing.T) {
+	tableName := fmt.Sprintf("cq_test_pk_rename_%d", rand.Intn(100))
```
Comment (erezrokah) [inline plugins/destination/postgresql/client/client_test.go:37]: This is so we don't need to drop the table before each test

Comment (disq) [reply plugins/destination/postgresql/client/client_test.go:38]: Might work or not work depending on random seed :)

Comment (erezrokah) [reply plugins/destination/postgresql/client/client_test.go:38]: Yeah, this is mostly for local dev, the CI always starts with a new DB

**Comment by yevgenypats** [comment]
I would be cautions in dropping things as a data-integration tool. users can (and do) create their own constraints, views and so on.

**Comment by erezrokah** [comment]
> I would be cautions in dropping things as a data-integration tool. users can (and do) create their own constraints, views and so on.

That's my concern too. We're dropping the constraint on columns that were primary keys before the migration, and are not after the migration.
Instead of dropping the constraint we can exit with an error guiding the user what to do.

The main issue is that the migration succeeds, but puts the DB in a state that any following sync fails. Exiting with an error should also work.

**Comment by yevgenypats** [comment]
> > I would be cautions in dropping things as a data-integration tool. users can (and do) create their own constraints, views and so on.
> 
> That's my concern too. We're dropping the constraint on columns that were primary keys before the migration, and are not after the migration. Instead of dropping the constraint we can exit with an error guiding the user what to do.
> 
> The main issue is that the migration succeeds, but puts the DB in a state that any following sync fails. Exiting with an error should also work.

Yes error would be a better/safer solution.

**Review by candiduslynx — APPROVED** [review]

**Comment by erezrokah** [comment]
> Yes error would be a better/safer solution.

Done in https://github.com/cloudquery/cloudquery/pull/6282/commits/14790b6f71ad3552a58d83c500406c392ed154d8

**Comment by disq** [comment]
Could also generate `drop not null` statements for the user to run.

**Comment by disq** [comment]
> Ah, I tried that at first. But the problem they only work after we drop the primary keys. And if we drop the primary keys we won't be able to detect the error the next time the migration runs

Need an `--allow-schema-change` option maybe? Which will auto-fix things by dropping PKs and remove constraints (and drop unused columns...)

For the user message you could also suggest the drop pk commands as well as drop not nulls.

**Comment by erezrokah** [comment]
> For the user message you could also suggest the drop pk commands as well as drop not nulls.

I felt that would be too verbose, but let me try that

**Comment by erezrokah** [comment]
> For the user message you could also suggest the drop pk commands as well as drop not nulls.

But then again, it brings the question - do we tell the user to drop the column? Drop only the constraint? Rename it?
The use case here is a primary key rename. We can't detect the rename just by looking at the DB

**Comment by disq** [comment]
> We can't detect the rename just by looking at the DB

In the olden times we did: I remember writing some code to compare a `schema.Table` with an actual table and to report what's changed and to generate sql migrations for the user. It detected column renames by [checking if the types were equal and the names were similar](https://github.com/cloudquery/cq-provider-sdk/blob/v0.9.5/migration/table.go#L213).

Here we'll just suggest the user to drop any unused column, including renames. We'll take care of the rest in the next migration (by adding what's missing). We could also tell the user what the next steps are (from our migrations POV, ie. what we're going to do on the next run if the columns are dropped), so they can decide for themselves if it's a rename and take necessary action? Too much? Maybe.

**Comment by erezrokah** [comment]
> In the olden times we did: I remember writing some code to compare a `schema.Table` with an actual table and to report what's changed and to generate sql migrations for the user. It detected column renames by [checking if the types were equal and the names were similar](https://github.com/cloudquery/cq-provider-sdk/blob/v0.9.5/migration/table.go#L213).
> 
> Here we'll just suggest the user to drop any unused column, including renames. We'll take care of the rest in the next migration (by adding what's missing). We could also tell the user what the next steps are (from our migrations POV, ie. what we're going to do on the next run if the columns are dropped), so they can decide for themselves if it's a rename and take necessary action? Too much? Maybe.

I could go either way. I'd rather give a generic error that's 100% correct, than have something wrong in our heuristics and provide the wrong recommendation.

Maybe dropping the primary key and the `not null` constraints is the safest option to recommend. I'll give it a try

**Comment by erezrokah** [comment]
image

OK, so I think https://github.com/cloudquery/cloudquery/pull/6282/commits/371be8e4533796fc8e09bf0b32ec76dd4addf744 should work (see image)

**Review by disq — APPROVED** [review]

**Pull request merged**